crossopt/assignment4/countislands.py

```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env python3
+
+
+"""grid of tiles. stores islands via dsu"""
+"""https://en.wikipedia.org/wiki/Disjoint-set_data_structure"""
+class Grid(object):
+    def __init__(self, n, m, grid):
+        self.n = n
+        self.m = m
+        self.grid = grid
+        self.parents = dict()
+        self.sizes = dict()
+        for i in range(n):
+            for j in range(m):
+                self.parents[i, j] = (i, j)
+                self.sizes[i, j] = 1
+                if i and self.grid[i][j] and self.grid[i - 1][j]:
+                    self.join((i, j), ((i - 1, j)))
+                if j and self.grid[i][j] and self.grid[i][j - 1]:
+                    self.join((i, j), ((i, j - 1)))
+
+    # returns unique id for each island
+    def __getitem__(self, u):
+        if self.parents[u] != u:
+            self.parents[u] = self[self.parents[u]]
+        return self.parents[u]
+    
+    def join(self, u, v):
```

Comment:
What is the complexity of join()?
Due to the recursive nature of __getitem__ it looks like it is O(n). 
I think this makes the  n^2. Am I right?

Reply:
With both path compression and union by size the complexity of join and get in DSU is O(alpha(n))